x-pack/plugins/lens/public/editor_frame_plugin/editor_frame.tsx

```diff
@@ -4,34 +4,139 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import React from 'react';
-import { Datasource, Visualization } from '../types';
+import React, { useState, useEffect } from 'react';
+import { Datasource, Visualization, DatasourcePublicAPI } from '../types';
 
 interface EditorFrameProps {
   datasources: { [key: string]: Datasource };
   visualizations: { [key: string]: Visualization };
+
+  activeDatasource: string | null;
+}
+
+interface DatasourceState {
+  [key: string]: {
+    state: any;
+    setState: (state: any) => void;
+    api: DatasourcePublicAPI;
+    datasource: Datasource;
+  };
+}
+
+interface VisualizationState {
+  [key: string]: {
+    state: any;
+    setState: (state: any) => void;
+    visualization: Visualization;
+  };
 }
 
 export function EditorFrame(props: EditorFrameProps) {
-  const keys = Object.keys(props.datasources);
+  const dsKeys = Object.keys(props.datasources);
+  const vKeys = Object.keys(props.visualizations);
+
+  const [datasourceState, setDatasourceState] = useState({} as DatasourceState);
+  const [visualizationState, setVisualizationState] = useState({} as VisualizationState);
+
+  // Initialize all datasources and load their public APIs
+  useEffect(
```

Comment:
Why initializing all datasources and visualizations? Is this for debugging purposes? If yes I would vote for adding dropdowns for switching between them because it's much closer to what this plugin will do in a finished product.